src/api/environment.cc

```diff
@@ -375,23 +375,33 @@ Environment* CreateEnvironment(
   // options than the global parse call.
   Environment* env = new Environment(
       isolate_data, context, args, exec_args, nullptr, flags, thread_id);
+
+  auto initialize_inspector = [&]() {
 #if HAVE_INSPECTOR
-  if (env->should_create_inspector()) {
-    if (inspector_parent_handle) {
-      env->InitializeInspector(
-          std::move(static_cast<InspectorParentHandleImpl*>(
-              inspector_parent_handle.get())->impl));
-    } else {
-      env->InitializeInspector({});
+    if (env->should_create_inspector()) {
+      if (inspector_parent_handle) {
+        env->InitializeInspector(
```

Comment:
Initializing the inspector when building the built-in snapshot used to not work because V8 used to create non-deserializable `v8::External` for certain console methods after we initialize the inspector (and invoked `v8_inspector::V8Inspector::contextCreated()` on the main context). In the case of user-land snapshot it's probably fine because it's reasonable to expect the user-land snapshot to be affected by the configuration/system state that the snapshot was built with, and we managed to [work around it with some code to clean the console up](https://github.com/nodejs/node/pull/44279) (it's probably not fail-safe though). In the case of the built-in snapshot we probably should just not initialize the inspector at all, since there's no guaranteed that a context that has gone through `v8_inspector::V8Inspector::contextCreated` is runtime-independent or deserializable (or add some debug options for that to be possible but turn that off in the default build).

Reply:
Also, are changes to the EnvironmentFlags considered breaking? If so it might be better to hide these extra flags somewhere else instead of keeping breaking the API for subtleties that only affect internals..

Reply:
> Also, are changes to the EnvironmentFlags considered breaking? 

So, this is the easy question: Not if they're purely additive and don't change default behavior. This patch isn't breaking our API or ABI for embedders, that I'm pretty sure of.

> In the case of the built-in snapshot we probably should just not initialize the inspector at all

This is a bit harder for me... when you say "built-in snapshot", you're referring to generating the snapshot, not consuming it, and you're specifically referring to `SnapshotMetadata::Type::kDefault`, right? In that case we already set the `kNoCreateInspector` flag in this PR, so I think we're already following this suggestion... right? 😅

Reply:
Oh, right, `if (env->should_create_inspector())` is wrapped inside `initialize_inspector()`. It might be easier to read if the conditions are just:

```c++
if (env->should_create_inspector() && env->should_create_inspector_before_bootstrap()) {
  initialize_inspector();
}
// .. bootstrap
if (env->should_create_inspector() && env->should_create_inspector_after_bootstrap()) {
  initialize_inspector();
}
```

Or we can also just expose `InitializeInspector()` and let the caller call it if it's necessary? Then there would not need to be an additional flag to control the timing of the inspector initialization (order-controlling flags might get out of control if there is another thing that needs to be done in a particular order later..)

Reply:
> Or we can also just expose `InitializeInspector()` and let the caller call it if `kNoCreateInspector` is set?

Yeah, I guess that’s the best thing to do then (we don’t even really need to expose `InitializeInspector()` for the purposes of this PR, so I’ll leave that out for now), see 6368ff01282a304bc3c0ea9457ca7adeed02a412. If we do want to give more fine-grained control over the behavior here, then we can always revisit this in the future.